lgn-worker/src/main.rs

```diff
@@ -114,12 +129,195 @@ fn main() -> anyhow::Result<()> {
         .with_http_listener(([0, 0, 0, 0], config.prometheus.port))
         .install()?;
 
-    run(&config)
+    if let Some(grpc_url) = &config.avs.gateway_grpc_url {
+        run_with_grpc(&config, grpc_url).await?;
+    } else {
+        tokio::task::block_in_place(move || -> Result<()> {
+            run(&config)?;
+            Ok(())
+        })?;
+    }
+
+    Ok(())
 }
 
-fn run(config: &Config) -> Result<()> {
-    info!("Version: {}", env!("CARGO_PKG_VERSION"));
-    let lagrange_wallet = match (
+async fn maybe_download_checksum(config: &Config) -> Result<()> {
+    if config.public_params.skip_checksum {
+        return Ok(());
+    }
+
+    // Fetch checksum file
+    // The checksum file can be generated in two ways.
+    // 1- Run the worker, and it will download and spit out the checksum on disk
+    // 2- Manually download the params then install with the checksums bin crate and run checksums -c -r zkmr_params -a BLAKE3
+    let checksum_url = &config.public_params.checksum_url;
+    let expected_checksums_file = &config.public_params.checksum_expected_local_path;
+
+    let response = reqwest::get(checksum_url)
+        .await
+        .context("Failed to fetch checksum file")?
+        .text()
+        .await
+        .context("Failed to read response text")?;
+
+    let mut file = tokio::fs::File::create(expected_checksums_file)
+        .await
+        .context("Failed to create local checksum file")?;
+
+    file.write_all(response.as_bytes())
+        .await
+        .context("Failed to write checksum file")?;
+
+    drop(file);
+
+    Ok(())
+}
+
+async fn run_with_grpc(config: &Config, grpc_url: &str) -> Result<()> {
+    let uri = grpc_url.parse::<tonic::transport::Uri>()?;
+    let (mut outbound, outbound_rx) = tokio::sync::mpsc::channel(1024);
+
+    info!("Verifying the checksums");
+
+    maybe_download_checksum(config).await?;
+
+    let mut provers_manager =
+        tokio::task::block_in_place(move || -> Result<ProversManager<TaskType, ReplyType>> {
```

Comment:
Why do we need the `block_in_place`?

Reply:
We are in the async context and we are about to do a huge blocking CPU-bound operation in the prover. Seems to be exactly the use-case for [`block_in_place`](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) ?

Reply:
But the prover won't have anything to do at that point, so there is nothing to block, right?

Reply:
Ah, I see, in this particular case the provers I believe are going to do blocking downloads with `reqwest::blocking::`

Reply:
And to give you some historical context, the only reason the worker is async-based is to handle websocket. Intrinsically, the worker is a simple fetch task/do work/send result, so there is virtually never two things happening concurrently.

Reply:
As much as I understand the sentiment it is getting harder and harder with the current Rust ecosystem not to pollute the code with async.

Reply:
Agree; but here the worker will not do anything in any case, so blocking the thread is not an issue.

Reply:
Yeah but let's future-proof this just in case.